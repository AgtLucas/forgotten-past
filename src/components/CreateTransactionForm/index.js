import React from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import Input from '../Input';
import Button from '../Button';
import ErrorMessage from '../ErrorMessage';
import Select from '../Select';
import styles from './styles.module.css';

const TransactionSchema = Yup.object().shape({
  description: Yup.string()
    .min(2, 'Descreva melhor sua transação.')
    .max(100, 'Calma aí, não precisa escrever 1 livro também...')
    .required('Você deve descrever a sua transação.'),
  price: Yup.number()
    .required('Por favor, adicione o valor da transação.'),
  transactionType: Yup.string()
    .required('Por favor, escolha o tipo da transação.')
});

const CreateTransactionForm = ({ handleSubmit }) => (
  <Formik
    initialValues={{ description: '', price: '', transactionType: '' }}
    validationSchema={TransactionSchema}
    onSubmit={(values, { setSubmitting, resetForm }) => {
      handleSubmit(values);
      resetForm();
      setSubmitting(false);
    }}
  >
    {({
      values,
      errors,
      touched,
      handleChange,
      handleBlur,
      handleSubmit,
      isSubmitting,
    }) => (
      <form
        method='post'
        className={styles.form}
        onSubmit={handleSubmit}
      >
        <label className={styles.label}>
          Tipo de transação
          <Select
            name='transactionType'
            value={values.transactionType}
            onChange={handleChange}
            onBlur={handleBlur}
          >
            <option>Escolha o tipo de transação</option>
            <option value='credit'>Crédito</option>
            <option value='debit'>Débito</option>
          </Select>
          {errors.transactionType && touched.transactionType ? (
            <ErrorMessage>{errors.transactionType}</ErrorMessage>
          ) : null}
        </label>
        <label className={styles.label}>
          Descrição
          <Input
            type='text'
            name='description'
            value={values.description}
            placeholder='Ex.: Fatura Banco XYZ'
            onChange={handleChange}
            onBlur={handleBlur}
          />
          {errors.description && touched.description ? (
            <ErrorMessage>{errors.description}</ErrorMessage>
          ) : null}
        </label>
        <label className={styles.label}>
          Valor
          <Input
            min='0'
            type='number'
            name='price'
            value={values.price}
            placeholder='Ex.: 200'
            onChange={handleChange}
            onBlur={handleBlur}
          />
          {errors.price && touched.price ? (
            <ErrorMessage>{errors.price}</ErrorMessage>
          ) : null}
        </label>
        <Button type='submit' disabled={isSubmitting}>Adicionar transação</Button>
      </form>
    )}
  </Formik>
);

export default CreateTransactionForm;