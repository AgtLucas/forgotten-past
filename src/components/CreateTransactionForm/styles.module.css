.form {
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  width: 400px;
  max-width: 100%;
  margin: 0 auto 50px;
  padding: 0 20px;
}

.label {
  width: 100%;
  height: 95px;
}

.button {
  margin-top: 10px;
}