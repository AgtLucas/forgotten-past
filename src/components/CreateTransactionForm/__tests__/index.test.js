import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CreateTransactionForm from '../';

describe('CreateTransactionForm', () => {
  it('should render CreateTransactionForm', () => {
    const props = {
      transaction: {
        description: 'Foo',
        price: 100,
        type: 'credit'
      },
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
    };

    const wrapper = shallow(<CreateTransactionForm {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});