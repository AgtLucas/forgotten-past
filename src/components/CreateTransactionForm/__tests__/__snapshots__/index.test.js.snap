// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTransactionForm should render CreateTransactionForm 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "description": "",
      "id": "",
      "price": "",
      "transactionType": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "transactionType",
        "price",
        "description",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "description": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "min": true,
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
        "price": NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
        "transactionType": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
>
  <Component />
</Formik>
`;
